### PR TITLE
Improve pills

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/16948
Fixes https://github.com/vector-im/element-web/issues/21281

<hr>

https://user-images.githubusercontent.com/25768714/126071702-4a19271e-effc-44a1-943e-36dcb29881d7.mp4

https://user-images.githubusercontent.com/25768714/126071703-3f8b7e89-f2fb-42b6-bc64-72059bba8991.mp4

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improve pills ([\#6398](https://github.com/matrix-org/matrix-react-sdk/pull/6398)). Fixes vector-im/element-web#16948 and vector-im/element-web#21281.

Preview: https://pr6398--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/views/elements/Pill.tsx [hunk 1]
```diff
@@ -28,56 +27,76 @@ import { Action } from "../../../dispatcher/actions";
...
+            group: null,
+            room: null,
+            hover: false,
+        };
+    }
 
     // TODO: [REACT-WARNING] Replace with appropriate lifecycle event
     // eslint-disable-next-line camelcase
-    async UNSAFE_componentWillReceiveProps(nextProps) {
+    async UNSAFE_componentWillReceiveProps(nextProps: IProps) {
```
Comment (SimonBrandner) [inline src/components/views/elements/Pill.tsx:99]: I don't think I see a reason to not just do this in `componentDidMount()`

**Comment by robintown** [comment]
Could the mini avatar inside space pills be made square-ish as well?

**Comment by SimonBrandner** [comment]
> Could the mini avatar inside space pills be made square-ish as well?

Why not? :)

File: src/components/views/elements/Pill.tsx (hunk 1 above)
Comment (SimonBrandner) [reply src/components/views/elements/Pill.tsx:97]: NVM, makes sense

File: res/css/_components.scss [hunk 2]
```diff
@@ -127,6 +127,7 @@
 @import "./views/elements/_ErrorBoundary.scss";
 @import "./views/elements/_EventListSummary.scss";
 @import "./views/elements/_FacePile.scss";
+@import "./views/elements/_Pill.scss";
```
Comment (turt2live) [inline res/css/_components.scss:130]: please use the reskindex script so this gets sorted into the correct spot

File: src/components/views/elements/Pill.tsx [hunk 3]
```diff
@@ -13,71 +13,88 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+    // ID/alias of the room/user
+    resourceId: string;
+    // Type of pill
+    pillType: string;
+    // The member related to the user pill
+    member: RoomMember;
+    // The group related to the group pill
+    group: Group;
+    // The room related to the room pill
+    room: Room;
```
Comment (turt2live) [inline src/components/views/elements/Pill.tsx:59]: these can probably be optional

File: src/components/views/elements/Pill.tsx [hunk 4]
```diff
@@ -13,71 +13,88 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
 import { replaceableComponent } from "../../../utils/replaceableComponent";
+import { Group } from "matrix-js-sdk/src/models/group";
+import { MatrixClient } from 'matrix-js-sdk/src';
+import RoomAvatar from '../avatars/RoomAvatar';
+import MemberAvatar from '../avatars/MemberAvatar';
+import BaseAvatar from '../avatars/BaseAvatar';
+
+interface IProps {
+    // The Type of this Pill. If url is given, this is auto-detected.
+    type?: string;
```
Comment (turt2live) [inline src/components/views/elements/Pill.tsx:38]: this is screaming for an enum

**Review by turt2live — COMMENTED** [review]
looks largely sane code-wise, but definitely needs design to take a look at it in more detail.

**Review by turt2live — APPROVED** [review]

**Comment by nadonomy** [comment]
Notes on each change separately..

**On https://github.com/vector-im/element-web/issues/17248 / circles to squares:**

We wouldn't want to merge this change. I can see the core intent but ultimately it's arbitrary, subjective and has far reaching consequences. @SimonBrandner would you be OK to roll these back to continue with this PR?

**On https://github.com/vector-im/element-web/issues/16948 / hover states:**

@gaelledel would you be ok to look a little closer at this? The hover states will likely feel great, but wanted to loop you in to take a closer look at colour values etc as you've been cleaning so much up so far!

Thanks for the contribution! 🎉

**Comment by ShadowJonathan** [comment]
@nadonomy could you please put the feedback on circles to squared also on the corresponding issue?

**Comment by SimonBrandner** [comment]
> **On [vector-im/element-web#17248](https://github.com/vector-im/element-web/issues/17248) / circles to squares:**
> 
> We wouldn't want to merge this change. I can see the core intent but ultimately it's arbitrary, subjective and has far reaching consequences. @SimonBrandner would you be OK to roll these back to continue with this PR?

Yeah, I agree, I don't feel about it very strongly either, have reverted the changes.

> **On [vector-im/element-web#16948](https://github.com/vector-im/element-web/issues/16948) / hover states:**
> 
> @gaelledel would you be ok to look a little closer at this? The hover states will likely feel great, but wanted to loop you in to take a closer look at colour values etc as you've been cleaning so much up so far!
 
Currently, this adds a completely new colour for the hover state, which isn't great at all. We're also using a colour that isn't in Compound for the regular non-hovered pills. It would be great to get all of this in line with the Figam spec.

> Thanks for the contribution! tada

No problem!

**Comment by nadonomy** [comment]
> @nadonomy could you please put the feedback on circles to squared also on the corresponding issue?
> 
> I appreciate it, but I think it's more helpful to track that feedback somewhere where it'll be easier to find in the future

Will do! Although, just to possibly 180 on my last comment— we found the different radii were compromising visual harmony. The lack of unity ends up being less satisfying. Just exploring if tweaking the implementation with some spacing/padding might do the trick.

**Comment by ShadowJonathan** [comment]
> Just exploring if tweaking the implementation with some spacing/padding might do the trick.

I'd suggest that'd happen on a different PR then, to land this one, and experiment in another one (possibly in a feedback cycle with design)

**Comment by SimonBrandner** [comment]
The preview might be a little broken due to the PR not having updates for a while, will fix later today

**Comment by SimonBrandner** [comment]
Preview should now be fine

**Review by gaelledel — COMMENTED** [review]
Can the pills always have the same radius? Sometimes it looks squary where it should ideally be rounded all the time

**Review by gaelledel — APPROVED** [review]

**Pull request merged**